arch/arm/arm64/include/uk/asm/paging.h

```diff
@@ -47,19 +65,213 @@ struct ukarch_pagetable {
+
 #define PT_LEVELS			4
+#define PT_PTES_PER_LEVEL		512
+#define PT_LEVEL_SHIFT			9
 
 /* We use plain values here so we do not create dependencies on external helper
- * macros, which would forbit us to use the macros in functions defined further
+ * macros, which would forbid us to use the macros in functions defined further
  * down in this header.
  */
 #define PAGE_LEVEL			0
 #define PAGE_SHIFT			12
 #define PAGE_SIZE			0x1000UL
 #define PAGE_MASK			(~(PAGE_SIZE - 1))
 
-/* TODO: Choose appropriate value */
-#define __VADDR_INV			0UL
-#define __PADDR_INV			0UL
+#define PAGE_ATTR_PROT_MASK		0x0f
+#define PAGE_ATTR_PROT_SHIFT		0
+
+#define PAGE_ATTR_PROT_NONE		0x00
+#define PAGE_ATTR_PROT_READ		0x01
+#define PAGE_ATTR_PROT_WRITE		0x02
+#define PAGE_ATTR_PROT_EXEC		0x04
+
+#define PAGE_ATTR_TYPE_MASK		0x07
+#define PAGE_ATTR_TYPE_SHIFT		5
+
+#define PAGE_ATTR_TYPE_NORMAL_WB	(0 << PAGE_ATTR_TYPE_SHIFT)
+#define PAGE_ATTR_TYPE_NORMAL_WT	(1 << PAGE_ATTR_TYPE_SHIFT)
+#define PAGE_ATTR_TYPE_NORMAL_NC	(2 << PAGE_ATTR_TYPE_SHIFT)
+#define PAGE_ATTR_TYPE_DEVICE_nGnRnE	(3 << PAGE_ATTR_TYPE_SHIFT)
+#define PAGE_ATTR_TYPE_DEVICE_nGnRE	(4 << PAGE_ATTR_TYPE_SHIFT)
+#define PAGE_ATTR_TYPE_DEVICE_GRE	(5 << PAGE_ATTR_TYPE_SHIFT)
+
+#define PAGE_ATTR_SHAREABLE_MASK	0x03
+#define PAGE_ATTR_SHAREABLE_SHIFT	8
+
+#define PAGE_ATTR_SHAREABLE_NS		(0 << PAGE_ATTR_SHAREABLE_MASK)
+#define PAGE_ATTR_SHAREABLE_IS		(1 << PAGE_ATTR_SHAREABLE_MASK)
+#define PAGE_ATTR_SHAREABLE_OS		(2 << PAGE_ATTR_SHAREABLE_MASK)
+
+#define ARM64_PADDR_BITS		48
+#define ARM64_VADDR_BITS		48
+
+#define PT_Lx_PTES(lvl)			PT_PTES_PER_LEVEL
+
+#define PT_Lx_PTE_PRESENT(pte, lvl) \
+	((pte & PTE_Lx_VALID_BIT))
+
+#define PT_Lx_PTE_CLEAR_PRESENT(pte, lvl) \
+	(pte & ~PTE_Lx_VALID_BIT)
+
+#define PT_MAP_LEVEL_MAX	(PT_LEVELS - 2)
+
+#define PAGE_Lx_HAS(lvl) ((lvl) <= PT_MAP_LEVEL_MAX)
+
+#define PAGE_Lx_IS(pte, lvl)				\
+	((lvl == PAGE_LEVEL) || ((pte) & PTE_Lx_TYPE_BIT) == PTE_Lx_TYPE_BLOCK)
+
+#define PAGE_Lx_SHIFT(lvl)				\
+	(PAGE_SHIFT + (PT_LEVEL_SHIFT * lvl))
+
+#define PT_Lx_IDX(vaddr, lvl)				\
+	(((vaddr) >> PAGE_Lx_SHIFT(lvl)) & (PT_PTES_PER_LEVEL - 1))
+
+/* Any address controlled by TTBR1, ie bit[55] == 1 */
+#define ARM64_INVALID_ADDR		0xBAADBAADBAADBAADUL
+
+#define __VADDR_INV					\
+	PAGE_Lx_ALIGN_DOWN(ARM64_INVALID_ADDR, PT_MAP_LEVEL_MAX)
+
+#define __PADDR_INV					\
+	PAGE_Lx_ALIGN_DOWN(ARM64_INVALID_ADDR, PT_MAP_LEVEL_MAX)
 
-/* TODO: Complement missing definitions */
+/* Any PTE with bit[0] == 0 is invalid */
+#define PT_Lx_PTE_INVALID(lvl)		0x0UL
+
+#define ARM64_PADDR_MAX ((1ULL << ARM64_PADDR_BITS) - 1)
+#define ARM64_VADDR_MAX ((1ULL << ARM64_VADDR_BITS) - 1)
+
+#define ARM64_PADDR_VALID(paddr) (paddr <= (__paddr_t)ARM64_PADDR_MAX)
+#define ARM64_VADDR_VALID(vaddr) (vaddr <= (__vaddr_t)ARM64_VADDR_MAX)
+
+#ifndef __ASSEMBLY__
+static inline __paddr_t PT_Lx_PTE_PADDR(__pte_t pte, unsigned int lvl)
+{
+	__paddr_t paddr;
+	__u64 pte_lx_map_paddr_mask[] = {
```

Comment:
You can achieve much better optimization of the function if you define this as `static`. This will also increase the chance of inlining 😃
```suggestion
	static __u64 pte_lx_map_paddr_mask[] = {
```
Not static:
![image](https://user-images.githubusercontent.com/8928900/182779539-2c5408e9-4af3-4444-a984-915f7f1b2623.png)

Static:
![image](https://user-images.githubusercontent.com/8928900/182779652-faea502d-fe08-4334-881f-0372b1ad6a0c.png)

Reply:
That's a cool optimization, thanks! 👍🏼